versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/RegLogSpliterator.java

```diff
@@ -0,0 +1,167 @@
+ * limitations under the License.
+ */
+package org.projectnessie.versioned.persist.nontx;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.Spliterator;
+import java.util.Spliterators.AbstractSpliterator;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.projectnessie.versioned.Hash;
+import org.projectnessie.versioned.RefLogNotFoundException;
+import org.projectnessie.versioned.persist.adapter.RefLog;
+
+/**
+ * {@link Spliterator} of {@link RefLog} entries using a variable number of ref-log stripes,
+ * according to {@link NonTransactionalDatabaseAdapterConfig#getRefLogStripes()} to read {@link
+ * RefLog}s from in parallel.
+ *
+ * <p>The next {@link RefLog} returned by this implementation is the next available from all stripes
+ * with the lowest {@link RefLog#getOperationTime()}.
+ */
+final class RegLogSpliterator extends AbstractSpliterator<RefLog> {
+
+  private final List<RefLogSplit> splits;
+
+  private final Hash initialHash;
+  private boolean initialHashSeen;
+  private RefLog initialRefLog;
+
+  RegLogSpliterator(Hash initialHash, Stream<Spliterator<RefLog>> refLogStripeFetcher)
+      throws RefLogNotFoundException {
+    super(Long.MAX_VALUE, 0);
+
+    this.initialHash = initialHash;
+    this.initialHashSeen = initialHash == null;
+
+    splits = refLogStripeFetcher.map(RefLogSplit::new).collect(Collectors.toList());
+
+    // The API for DatabaseAdapter.refLog(Hash) requires to throw a RefLogNotFoundException,
+    // if no RefLog for the initial hash exists. This loop tries to find the initial RefLog entry
+    // and throw
+    AtomicReference<RefLog> initial = new AtomicReference<>();
+    while (!initialHashSeen) {
+      if (!tryAdvance(initial::set)) {
+        break;
+      }
+    }
+
+    // Throw RefLogNotFoundException, if RefLog with initial hash could not be found.
+    if (!initialHashSeen) {
+      throw RefLogNotFoundException.forRefLogId(initialHash.asString());
+    }
+    this.initialRefLog = initial.get();
+  }
+
+  @Override
+  public boolean tryAdvance(Consumer<? super RefLog> action) {
+    if (initialRefLog != null) {
+      // This returns the RefLog entry for the initial hash.
+      action.accept(initialRefLog);
+      initialRefLog = null;
+      return true;
+    }
+
+    Optional<RefLogSplit> oldest =
+        splits.stream()
+            .filter(RefLogSplit::hasMore)
+            .max(Comparator.comparing(RefLogSplit::operationTime));
+
+    if (!oldest.isPresent()) {
+      return false;
+    }
+
+    RefLog refLog = oldest.get().pull();
+    if (refLog == null) {
+      return false;
+    }
+
+    if (!initialHashSeen) {
+      initialHashSeen = refLog.getRefLogId().equals(initialHash);
+    }
+
+    if (initialHashSeen) {
+      action.accept(refLog);
+    }
+
+    return true;
```

Comment:
Returning `true` without calling the `action` appears to violate the javadoc contract of the `tryAdvance` method. If we reversed `if/while` conditions on lines 60-61 we could call the action unconditionally, I guess.